### ISSUE TITLE
Pending pvc

### ISSUE THREAD

**Issue by hcs129** [kubernetes/kubernetes#50522]
I tried to setup a pvc using my storageclasses, it is at pending state and the output is as follows:

> Name: testclaim
> Namespace: default
> StorageClass: fastrbd
> Status: Pending
> Volume:
> Labels: 
> Annotations: kubectl.kubernetes.io/last-applied-configuration={"apiVersion":"v1","kind":"PersistentVolumeClaim","metadata":{"annotations":{},"name":"testclaim","namespace":"default"},"spec":{"accessModes":["ReadWr...
> volume.beta.kubernetes.io/storage-class=fastrbd
> volume.beta.kubernetes.io/storage-provisioner=kubernetes.io/rbd
> Capacity:
> Access Modes:
> Events:
> FirstSeen LastSeen Count From SubObjectPath Type Reason Message
> 
> 11s 9s 2 persistentvolume-controller Warning ProvisioningFailed Failed to provision volume with StorageClass "fastrbd": rbd: create volume failed, err: executable file not found in $PATH

Do you have any suggestion?

**Comment by zhangxiaoyu-zidif** [comment]
'''yaml
Name: testclaim
Namespace: default
StorageClass: fastrbd
Status: Pending
Volume:
Labels:
Annotations: kubectl.kubernetes.io/last-applied-configuration={"apiVersion":"v1","kind":"PersistentVolumeClaim","metadata":{"annotations":{},"name":"testclaim","namespace":"default"},"spec":{"accessModes":["ReadWr...
volume.beta.kubernetes.io/storage-class=fastrbd
volume.beta.kubernetes.io/storage-provisioner=kubernetes.io/rbd
Capacity:
Access Modes:
Events:
FirstSeen LastSeen Count From SubObjectPath Type Reason Message
'''

**Comment by wongma7** [comment]
`rbd` binary must be installed on master & nodes e.g. via ceph-common package

**Comment by lenguyen89** [comment]
@wongma7 
Can you please elaborate a little bit on this?

> rbd binary must be installed on master & nodes e.g. via ceph-common package

Do I just do apt-get ceph-common and that's it? Do I need to restart kubernetes or run some command to let it know about the rbd package?

**Comment by wongma7** [comment]
@lenguyen89 yes you should just install and that's it, basically kubernetes is trying to call out to `rbd ...` so as soon as that rbd binary exists in PATH, things will start working

**Comment by lenguyen89** [comment]
@wongma7 
Thanks for the speedy reply!
My pvc now give me this error

`2017-09-01 03:12:51.423305 7fc61c9d6700  0 -- 10.12.1.1:0/1007848 >> 10.12.1.14:6789/0 pipe(0x7fc61400fb30 sd=4 :36134 s=1 pgs=0 cs=0 l=1 c=0x7fc61400a070).fault
2017-09-01 03:12:51.424179 7fc61c9d6700  0 -- 10.12.1.1:0/1007848 >> 10.12.1.14:6789/0 pipe(0x7fc61400fb30 sd=3 :36136 s=1 pgs=0 cs=0 l=1 c=0x7fc61400a070).connect protocol feature mismatch, my fffffffffff < peer 4010ff8eea4fffb missing 401000000000000
2017-09-01 03:12:51.625366 7fc61c9d6700  0 -- 10.12.1.1:0/1007848 >> 10.12.1.14:6789/0 pipe(0x7fc61400fb30 sd=3 :36138 s=1 pgs=0 cs=0 l=1 c=0x7fc61400a070).connect protocol feature mismatch, my fffffffffff < peer 4010ff8eea4fffb missing 401000000000000
2017-09-01 03:12:52.026627 7fc61c9d6700  0 -- 10.12.1.1:0/1007848 >> 10.12.1.14:6789/0 pipe(0x7fc61400fb30 sd=3 :36140 s=1 pgs=0 cs=0 l=1 c=0x7fc61400a070).connect protocol feature mismatch, my fffffffffff < peer 4010ff8eea4fffb missing 401000000000000
2017-09-01 03:12:52.827941 7fc61c9d6700  0 -- 10.12.1.1:0/1007848 >> 10.12.1.14:6789/0 pipe(0x7fc61400fb30 sd=3 :36142 s=1 pgs=0 cs=0 l=1 c=0x7fc61400a070).connect protocol feature mismatch, my fffffffffff < peer 4010ff8eea4fffb missing 401000000000000
2017-09-01 03:12:54.395572 7fc624029780  0 monclient(hunting): authenticate timed out after 300
2017-09-01 03:12:54.395618 7fc624029780  0 librados: client.admin authentication error (110) Connection timed out
rbd: couldn't connect to the cluster!`

I did a little bit of googling and it's looked like that the client is missing 
CEPH_FEATURE_CRUSH_TUNABLES5 and CEPH_FEATURE_NEW_OSDOPREPLY_ENCODING

`my fffffffffff < peer 4010ff8eea4fffb missing 401000000000000`
http://cephnotes.ksperis.com/blog/2014/01/21/feature-set-mismatch-error-on-ceph-kernel-client

I made sure that my client and cluster version match up. They are both
`ceph version 12.2.0 (32ce2a3ae5239ee33d6150705cdb24d43bab910c) luminous (rc)`

Can you give me some pointer on this problem? Like is there a place where I can see all kubernetes internal volume plugin and their versions?

**Comment by wongma7** [comment]
Sorry, I don't know about that. I would open a new issue

**Comment by hcs129** [comment]
@wongma7 
Thanks for you reply. I realized the missing of ceph-common.
Do you have any idea how to install it on the minikube virtual nodes?

**Comment by wongma7** [comment]
no, I'm not sure what OS minikube is under the hood. You may have to use some kind of containerized ceph utilities https://github.com/rook/rook/blob/master/Documentation/k8s-pre-reqs.md#ceph-and-rbd-utilities-installed-on-the-nodes

**Comment by somejfn** [comment]
Same problem here on Tectonic 1.7.2 and Ceph Luminous 12.2.0

W0918 17:49:34.813572       1 rbd_util.go:364] failed to create rbd image, output 2017-09-18 17:44:34.641191 7fe33b924780 -1 did not load config file, using default settings.
2017-09-18 17:44:34.806063 7fe33b91c700  0 -- 10.74.80.70:0/1000057 >> 10.74.80.131:6789/0 pipe(0x55e354476060 sd=3 :32796 s=1 pgs=0 cs=0 l=1 c=0x55e3544762f0).connect protocol feature mismatch, my fffffffffff < peer 4010ff8eea4fffb missing 401000000000000

But afaik since the rbd utility is bundled inside the _hyperkube_ image,  isn;t that image that needs updating ?

**Comment by nesc58** [comment]
I want to use ceph 12.x but I cannot create persist volume claims. The log of the kubernetes-controll-manager contains a lot of this lines:
```W0927 06:24:21.432175       1 rbd_util.go:395] failed to create rbd image, output 2017-09-27 06:24:21.422704 7fc501794780 -1 did not load config file, using default settings.
rbd: extraneous parameter --image-feature```

The rbd version in the hyperkube container is 0.80.7
```
# rbd --version
ceph version 0.80.7 (6c0127fcb58008793d3c8b62d925bc91963672a3)
```

My setup: CoreOS (1465.8.0)  + Kubernetes (Hyperkube version 1.7.x / and also tried 1.8.x)

**Comment by hzamani** [comment]
Adding `imageFormat: "1"` to `storageclass.parameters` solved my problem on `v1.8.1+coreos.0`. Another solution is to build your own image based on hyperkube image.

**Comment by tanhui2333** [comment]
I have same problem
that is output:
Events:
  FirstSeen	LastSeen	Count	From				SubObjectPath	Type		Reason			Message
  ---------	--------	-----	----				-------------	--------	------			-------
  41m		8m		133	persistentvolume-controller		Warning		ProvisioningFailed	Failed to provision volume with StorageClass "ceph-rbd": failed to create rbd image: executable file not found in $PATH, command output:
  4m		5s		20	persistentvolume-controller		Warning		ProvisioningFailed	Failed to provision volume with StorageClass "ceph-rbd": failed to create rbd image: executable file not found in $PATH, command output:

**Comment by JrCs** [comment]
Is it possible to upgrade the ceph-common library in the hyperkube image to a more recent version ?

**Comment by somejfn** [comment]
This is what https://github.com/kubernetes/kubernetes/issues/45010 is for.  Tho it wont bring ceph-common to Luminous release.

**Comment by f7r** [comment]
Same problem as @lenguyen89
I use Kubernetes v1.8.2, deploy by kubespray on CentOS 7.4
Ceph version is luminous(12.2.1)

describe pvc output as below:

    2017-11-10 09:16:20.569530 7efd702f6700  0 -- 172.22.117.221:0/1001521 >> 172.22.117.221:6789/0 pipe(0x7efd5c00e650 sd=3 :33126 s=1 pgs=0 cs=0 l=1 c=0x7efd5c009120).connect protocol feature mismatch, my fffffffffff < peer 4010ff8eea4fffb missing 401000000000000
    2017-11-10 09:16:21.371104 7efd702f6700  0 -- 172.22.117.221:0/1001521 >> 172.22.117.221:6789/0 pipe(0x7efd5c00e650 sd=3 :33128 s=1 pgs=0 cs=0 l=1 c=0x7efd5c009120).connect protocol feature mismatch, my fffffffffff < peer 4010ff8eea4fffb missing 401000000000000
    2017-11-10 09:16:22.904721 7efd75945780  0 monclient(hunting): authenticate timed out after 300
    2017-11-10 09:16:22.904874 7efd75945780  0 librados: client.admin authentication error (110) Connection timed out
    rbd: couldn't connect to the cluster!

**Comment by f7r** [comment]
If I set imageFormat: "2", and imageFeatures: layering in storageclass.parameters,
The problem will same as @nesc58 

describe pvc output as below:

    Events:
      Type     Reason              Age   From                         Message
      ----     ------              ----  ----                         -------
      Warning  ProvisioningFailed  14s   persistentvolume-controller  Failed to provision volume with StorageClass "rbd-storage": failed to create rbd image: exit status 1, command output: 2017-11-10 09:30:45.300114 7f7da8e8f780 -1 did not load config file, using default settings.
    rbd: extraneous parameter --image-feature

**Comment by discostur** [comment]
I created a new kube-controller-image with debian stretch, kube 1.8.4 and ceph 12.2.1 ... k8s nodes are running centos 7; imageFormat: 1

Still seeing the following error:

``` 
protocol feature mismatch, missing 400000000000000
``` 

Any Ideas?

**Comment by alv91** [comment]
For me fixed with kubernetes 1.9.0
```
rbd --version
ceph version 10.2.5 (c461ee19ecbc0c5c330aca20f7392c9a00730367)
```
